STM32 OSPI NOR: run application from external flash memory

FRASTM opened the pull request:
Following the [ stm32h5 run application in external flash memory XIP](https://github.com/zephyrproject-rtos/zephyr/pull/88579)

This PR makes the **samples/sysbuild/with_mcuboot --sysbuild**  running on a variant of the  stm32u585 disco kit.

Requires the ./bootloader/mcuboot/bootzephyr/boards/b_u585i_iot02a_stm32u585xx_ext_flash_app.conf : 
```
CONFIG_BOOT_DIRECT_XIP=y
CONFIG_STM32_MEMMAP=y
```


run the `west build -p -b b_u585i_iot02a/stm32u585xx/ext_flash_app samples/sysbuild/with_mcuboot --sysbuild`

```
*** Booting MCUboot v2.1.0-rc1-390-gfbc9aff5acc1 ***
*** Using Zephyr OS build v4.2.0-rc2-72-g8c58be376223 ***
*** Booting Zephyr OS build v4.2.0-rc2-72-g8c58be376223 ***
Address of sample 0x70000000
Hello sysbuild with mcuboot! b_u585i_iot02a
```


Requires boot/zephyr/boards/b_u585i_iot02a.conf
```
CONFIG_BOOT_DIRECT_XIP=y
CONFIG_STM32_MEMMAP=y
```

etienne-lms commented on boards/st/b_u585i_iot02a/board.cmake:
```diff
@@ -13,11 +13,9 @@ if(CONFIG_BUILD_WITH_TFM)
 
 # keep first
-if(CONFIG_STM32_MEMMAP)
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
```
Intentionally removing `--ersase` directive?
If so, could you mention it in the commit message?

FRASTM replied:
not necessarily, restored

FRASTM committed:
samples: sysbuild: with mcu_boot on the stm32u5 with octoSPI

Run the sample to execute in place on the external flash of the
b_u585i_iot02a variant 'ext_flash_app' disco kit,
with_boot application is built and linked for the external flash

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Define the partition of the external octo NOR in the b_u858i_iot02a.dts to avoid conflit with the partitioning of the b_u858i_iot02a_ns.dts.

FRASTM commented:
fixing compliance issues

FRASTM committed:
boards: st: b_u585i disco kit variant "ext_flash_app" for XIP

Create a new variant to execute in place on the external octoSPI
flash NOR of the b_u585i_iot02a disco kit
The flash size in the yaml corresponds to the the slot0
of that external NOR memory.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Creating a variant (only) for the b_u585i_iot02a disco kit : b_u585i_iot02a/stm32u585xx/xip 
Requires the mcuboot to define the corresponding conf file: ../bootloader/mcuboot/boot/zephyr/boards/b_u585i_iot02a_stm32u585xx_xip.conf
```
*** Booting MCUboot v2.1.0-rc1-390-gfbc9aff5acc1 ***
*** Using Zephyr OS build v4.2.0-rc3-36-ga08697fb1810 ***
*** Booting Zephyr OS build v4.2.0-rc3-38-g84c417194c74 ***
Address of sample 0x70000000
Hello sysbuild with mcuboot! b_u585i_iot02a
```

FRASTM commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_xip.dts:
```diff
@@ -0,0 +1,110 @@
+			reg = <0x00000000 DT_SIZE_K(416)>;
+		};
+
+		slot1_partition: partition@78000 {
```
fixed

FRASTM commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_xip.dts:
```diff
@@ -0,0 +1,110 @@
+
+		scratch_partition: partition@d0000 {
+			label = "image-scratch";
+			reg = <0x000d0000 DT_SIZE_K(128)>;
```
no particular reason, increasing to 192K

FRASTM commented:
Rebase on 47b07e5a09efdec59247770b5563b75b4ff75324
Add the link to mcuboot : https://github.com/mcu-tools/mcuboot/pull/2399

FRASTM commented:
waiting for the https://github.com/zephyrproject-rtos/zephyr/pull/93762 to report all modifications to stm32 ospi flash and then stm32 flash qspi drivers

FRASTM commented:
Rebase on 9354e064220310b691021bc69e4fff8b531b5020
Includes the https://github.com/zephyrproject-rtos/mcuboot/pull/137 now that PR has been  merged in the https://github.com/mcu-tools/mcuboot.git

FRASTM commented:
rebase on 727c15a038760774b992bf9c9e87f0dc793db224

etienne-lms commented on boards/st/b_u585i_iot02a/b_u585i_iot02a.dts:
```diff
@@ -112,3 +112,15 @@
+			reg = <0x00000000 DT_SIZE_M(64)>;
+		};
+	};
+};
```
This is already set at lines 79..89.
It was brought by 49ed8318c296217b13d320d97d72f0f99f657d41.

etienne-lms commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_ext_flash_app.yaml:
```diff
@@ -0,0 +1,10 @@
+  - zephyr
+  - gnuarmemb
+ram: 786
+flash: 512
```
416? (`slot0_partition` and `slot1_partition` sizes)

etienne-lms replied:
`flash:` tag allows twister to know if the board can be used for a samples/tests for zephyr image size contraints.
It should reflect the size of the chosen `zephyr,flash` partition IUUC.

FRASTM replied:
updated to "416"

etienne-lms commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,110 @@
+		zephyr,console = &usart1;
+		zephyr,shell-uart = &usart1;
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
```
What is the purpose of `zephyr,flash` here?

FRASTM replied:
Removing  
```
chosen {
		zephyr,flash = &flash0;
}
```

and the samples/hello_world/ cannot build anymore
```
section `rom_start' will not fit in region `FLASH'
 region `FLASH' overflowed by 17616 bytes
```

 It should though without any  mcuboot

etienne-lms commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,110 @@
+		reg = <0x70000000 DT_SIZE_M(64)>;
+		zephyr,memory-region = "EXTMEM";
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
+		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_IO) )>;
```
```suggestion
		zephyr,memory-attr = <DT_MEM_ARM(ATTR_MPU_IO)>;

etienne-lms commented on boards/st/b_u585i_iot02a/b_u585i_iot02a_stm32u585xx_ext_flash_app.yaml:
```diff
@@ -0,0 +1,10 @@
+identifier: b_u585i_iot02a/stm32u585xx/xip
```
```suggestion
b_u585i_iot02a/stm32u585xx/ext_flash_app
```
(issue reported by CI build tests)

FRASTM replied:
done

JarmouniA reviewed (changes requested):
Missing `Kconfig.sysbuild` at the board level, which shall contain at minimum
```
choice BOOTLOADER
	default BOOTLOADER_MCUBOOT
endchoice

choice MCUBOOT_MODE
	default MCUBOOT_MODE_DIRECT_XIP
endchoice
```

FRASTM commented:
to be refined according to the https://github.com/zephyrproject-rtos/zephyr/pull/97037
